lib/rucio/tests/test_judge_evaluator.py

```diff
@@ -143,11 +146,57 @@ def test_judge_add_dataset_to_container(self):
 
         # Check if the Locks are created properly
         for file in files:
-            assert(len(get_replica_locks(scope=file['scope'], name=file['name'])) == 2)
+            assert len(get_replica_locks(scope=file['scope'], name=file['name'])) == 2
 
         # Check if the DatasetLocks are created properly
-        dataset_locks = [lock for lock in get_dataset_locks(scope=scope, name=dataset)]
-        assert(len(dataset_locks) == 2)
+        dataset_locks = list(get_dataset_locks(scope=scope, name=dataset))
+        assert len(dataset_locks) == 2
+
+    @pytest.mark.noparallel(reason="uses mock scope and predefined RSEs; runs judge evaluator")
+    def test_judge_dataset_grouping_all(self):
+        """ JUDGE EVALUATOR: Test the judge when adding dataset to existing container with ALL grouping"""
+
+        # create a container
+        scope = InternalScope('mock', **self.vo)
+        parent_container = 'container_' + str(uuid())
+        add_did(scope, parent_container, DIDType.CONTAINER, self.jdoe)
+
+        # create a dataset, populate it with an "existing" file and declare that they reside in the T1 RSE
+        files = create_files(1, scope, self.rse1_id)        # rse1 has T1 tag
+        dataset1 = 'dataset_' + str(uuid())
+        add_did(scope, dataset1, DIDType.DATASET, self.jdoe)
+        attach_dids(scope, dataset1, files, self.jdoe)
+
+        # attach the dataset to the container
+        attach_dids(scope, parent_container, [{'scope': scope, 'name': dataset1}], self.jdoe)
+
+        # add rule to copy everything in this container to T1, use grouping=ALL
+        add_rule(dids=[{'scope': scope, 'name': parent_container}],
+                 account=self.jdoe,
+                 copies=1, rse_expression=self.T1, grouping='ALL',
+                 weight=None, lifetime=None, locked=False, subscription_id=None
```

Comment:
As currently written, I think the test will be non-deterministic as there is a 1 in 3 chance the RSESelector will select the same RSE as previously used. Perhaps a way to make it more likely to fail is to assign a weight here such that it almost certainly goes to a specific T1 RSEs, say `MOCK1`, then update the weight after L180 to be almost sure that `MOCK3` would be chosen when the re-evaluator runs next unless `preferred_rse_ids` is properly set to `MOCK1`.

Reply:
I don't think so. The rule here will always use the `rse1` since the data is already there. Since the new dataset is on `rse4` this should always be fine.

Reply:
Yes for this line it will use `rse1`, but the issue I am trying to trigger happens during the RSE selection process in L192 where the re-evaluator runs now with the new child `dataset2`. There we hope that the selection process puts locks on `dataset2` at `rse1` but in practice what I observe in production is it may sometimes put locks at `rse3`. In this test, if the grouping is not respected it still might select `rse1` by chance. Hence why I was advocating a weight in the rule.

Reply:
I see what you mean. 👍 
Currently there is two things going wrong here, which Igor already identified.
1. The re-evaluation of a dataset/container being added to a container at all. This might not be triggered.
2. The correct identification of the RSEs already selected.
This test will correctly test for 1, but not necessarily for 2.